File: Kernel Simulator/Shell/Shells/JsonShell.vb
﻿
'    Kernel Simulator  Copyright (C) 2018-2022  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.IO
Imports KS.Misc.Editors.JsonShell

Namespace Shell.Shells
    Public Class JsonShell
        Inherits ShellExecutor
        Implements IShell

        Public Overrides ReadOnly Property ShellType As ShellType Implements IShell.ShellType
            Get
                Return ShellType.JsonShell
            End Get
        End Property

        Public Overrides Property Bail As Boolean Implements IShell.Bail

        Public Overrides Sub InitializeShell(ParamArray ShellArgs() As Object) Implements IShell.InitializeShell
            'Get file path
            Dim FilePath As String = ""
            If ShellArgs.Length > 0 Then
                FilePath = ShellArgs(0)
            Else
                Write(DoTranslation("File not specified. Exiting shell..."), True, ColTypes.Error)
                Bail = True
            End If

            While Not Bail
                'Open file if not open
                If JsonShell_FileStream Is Nothing Then
                    Wdbg(DebugLevel.W, "File not open yet. Trying to open {0}...", FilePath)
                    If Not JsonShell_OpenJsonFile(FilePath) Then
                        Write(DoTranslation("Failed to open file. Exiting shell..."), True, ColTypes.Error)
                        Exit While
                    End If
                    JsonShell_AutoSave.Start()
                End If

                'See UESHShell.vb for more info
                SyncLock GetCancelSyncLock(ShellType)
                    'Prepare for prompt
                    If DefConsoleOut IsNot Nothing Then
                        Console.SetOut(DefConsoleOut)
                    End If
                    Wdbg(DebugLevel.I, "JsonShell_PromptStyle = {0}", JsonShell_PromptStyle)
                    If JsonShell_PromptStyle = "" Then
                        Write("[", False, ColTypes.Gray) : Write("{0}{1}", False, ColTypes.UserName, Path.GetFileName(FilePath), If(JsonShell_WasJsonEdited(), "*", "")) : Write("] > ", False, ColTypes.Gray)
                    Else
                        Dim ParsedPromptStyle As String = ProbePlaces(JsonShell_PromptStyle)
                        Conversion.ConvertVTSequences(ParsedPromptStyle)
                        Write(ParsedPromptStyle, False, ColTypes.Gray)
                    End If
                    SetInputColor()

                    'Raise the event
                    KernelEventManager.RaiseJsonShellInitialized()
                End SyncLock

                'Prompt for command
                Dim WrittenCommand As String = ReadLine()
                If Not (WrittenCommand = Nothing Or WrittenCommand?.StartsWithAnyOf({" ", "#"})) Then
                    KernelEventManager.RaiseJsonPreExecuteCommand(WrittenCommand)
                    GetLine(WrittenCommand, False, "", ShellType.JsonShell)
                    KernelEventManager.RaiseJsonPostExecuteCommand(WrittenCommand)
                End If
            End While

            'Close file
            JsonShell_CloseTextFile()
            JsonShell_AutoSave.Stop()
        End Sub

    End Class
End Namespace

File: public/Kernel Simulator/Shell/Shells/Json/JsonShell.cs
﻿
// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System;
using Extensification.StringExts;
using KS.ConsoleBase.Colors;
using KS.ConsoleBase.Inputs;
using KS.Kernel.Debugging;
using KS.Languages;
using KS.Misc.Editors.JsonShell;
using KS.Misc.Writers.ConsoleWriters;
using KS.Shell.Prompts;
using KS.Shell.ShellBase.Commands;
using KS.Shell.ShellBase.Shells;

namespace KS.Shell.Shells.Json
{
    /// <summary>
    /// The JSON editor shell
    /// </summary>
    public class JsonShell : ShellExecutor, IShell
    {

        /// <inheritdoc/>
        public override ShellType ShellType => ShellType.JsonShell;

        /// <inheritdoc/>
        public override bool Bail { get; set; }

        /// <inheritdoc/>
        public override void InitializeShell(params object[] ShellArgs)
        {
            // Get file path
            string FilePath = "";
            if (ShellArgs.Length > 0)
            {
                FilePath = Convert.ToString(ShellArgs[0]);
            }
            else
            {
                TextWriterColor.Write(Translate.DoTranslation("File not specified. Exiting shell..."), true, ColorTools.ColTypes.Error);
                Bail = true;
            }

            while (!Bail)
            {
                // Open file if not open
                if (JsonShellCommon.JsonShell_FileStream is null)
                {
                    DebugWriter.WriteDebug(DebugLevel.W, "File not open yet. Trying to open {0}...", FilePath);
                    if (!JsonTools.JsonShell_OpenJsonFile(FilePath))
                    {
                        TextWriterColor.Write(Translate.DoTranslation("Failed to open file. Exiting shell..."), true, ColorTools.ColTypes.Error);
                        break;
                    }
                    JsonShellCommon.JsonShell_AutoSave.Start();
                }

                // See UESHShell.cs for more info
                lock (CancellationHandlers.GetCancelSyncLock(ShellType))
                {
                    // Prepare for prompt
                    PromptPresetManager.WriteShellPrompt(ShellType);

                    // Raise the event
                    Kernel.Kernel.KernelEventManager.RaiseJsonShellInitialized();
                }

                // Prompt for command
                string WrittenCommand = Input.ReadLine();
                if ((string.IsNullOrEmpty(WrittenCommand) | (WrittenCommand?.StartsWithAnyOf(new[] { " ", "#" }))) == false)
                {
                    Kernel.Kernel.KernelEventManager.RaiseJsonPreExecuteCommand(WrittenCommand);
                    Shell.GetLine(WrittenCommand, "", ShellType.JsonShell);
                    Kernel.Kernel.KernelEventManager.RaiseJsonPostExecuteCommand(WrittenCommand);
                }
            }

            // Close file
            JsonTools.JsonShell_CloseTextFile();
            JsonShellCommon.JsonShell_AutoSave.Stop();
        }

    }
}
